From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Wed, 22 May 2024 10:12:04 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d0dc42bc30e81b6d7283dc1b3a42cab55fecafc7..f11dbf306e8b74352bd092093268b5117178a9fb 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3027,4 +3027,10 @@ public final class Bukkit {
         return server.getPhotographerManager();
     }
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    public static org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufManager() {
+        return server.getBytebufAllocator();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 4b37b44ff78503d4a1de808f5d429605f0de289c..6e9a0f48147e855e646223fedce3908548517f3e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2789,4 +2789,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Photographer API
     @NotNull org.leavesmc.leaves.entity.photographer.PhotographerManager getPhotographerManager();
     // Leaves end - Photographer API
+
+    // Leaves start - Bytebuf API
+    org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator();
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index e5947f299937a8100fd8c6b1456da9f9003c6b53..542e7f2deaacbc3a17a239154e55097111c5c4c3 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -71,7 +71,7 @@ import org.jspecify.annotations.Nullable;
  * Represents a player, connected or not
  */
 @NullMarked
-public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, ObjectContentsLike { // Paper
+public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient, ObjectContentsLike, org.leavesmc.leaves.bytebuf.PacketAudience { // Paper // Leaves - Bytbuf API
 
     // Paper start
     @Override
