From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 18 May 2025 19:59:11 +0800
Subject: [PATCH] Leaves Config API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index f11dbf306e8b74352bd092093268b5117178a9fb..ada86f4d598ac6774eabfd745f36bb759ad3a4ac 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3033,4 +3033,10 @@ public final class Bukkit {
         return server.getBytebufAllocator();
     }
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    public static org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig() {
+        return server.getLeavesConfig();
+    }
+    // Leaves end - Config API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6e9a0f48147e855e646223fedce3908548517f3e..1359fb4055f1995aeea78bdd285b08b2524fa73b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2793,4 +2793,8 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     // Leaves start - Bytebuf API
     org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator();
     // Leaves end - Bytebuf API
+
+    // Leaves start - Config API
+    org.leavesmc.leaves.config.LeavesConfigProvider getLeavesConfig();
+    // Leaves End - Config API
 }
