From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MC_XiaoHei <xor7xiaohei@gmail.com>
Date: Sat, 2 Aug 2025 09:46:11 +0800
Subject: [PATCH] Leaves Plugin

This patch is licensed under the MIT license.

diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index ada86f4d598ac6774eabfd745f36bb759ad3a4ac..52c78b1d12ff49777eb2eb89b6c5db4f668313af 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3011,6 +3011,12 @@ public final class Bukkit {
         server.restart();
     }
 
+    // Leaves start - Feature API
+    public static org.leavesmc.leaves.plugin.FeatureManager getFeatureManager() {
+        return server.getFeatureManager();
+    }
+    // Leaves end - Feature API
+
     // Leaves start - Bot API
     /**
      * Returns a bot manager.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1359fb4055f1995aeea78bdd285b08b2524fa73b..7138a624c636c5d29d809534f560f67866507545 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2777,6 +2777,10 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
 
+    // Leaves start - Feature API
+    org.leavesmc.leaves.plugin.FeatureManager getFeatureManager();
+    // Leaves End - Feature API
+
     // Leaves start - Bot API
     /**
      * Returns a bot manager.
