From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 30 Jul 2023 12:20:16 +0800
Subject: [PATCH] Zero tick plants


diff --git a/net/minecraft/world/level/block/BambooStalkBlock.java b/net/minecraft/world/level/block/BambooStalkBlock.java
index cf23ecd34bc7af694a036e388f20c56b0ca0a5b1..8c22957926e23a97b959f36bbe350db24a967736 100644
--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -108,6 +108,10 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index 1fbdb7a9ae8120cbe43cb185ad17c4769cf6e8b6..5b6124a1185e8a3b51f0e13777a4bc9bd75ae01b 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -47,6 +47,10 @@ public class CactusBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/ChorusFlowerBlock.java b/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 76dbad4969f8e7e39c78eec782579362203ca6bc..386795ed3920dc94341848b65dd2b29ab1a43cd1 100644
--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -47,6 +47,10 @@ public class ChorusFlowerBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
diff --git a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
index 21524350cb8822d33c86f1be783669ac8b961319..fbed46f137c13ae80a954d1de3b7b339978024ce 100644
--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -160,4 +160,15 @@ public abstract class GrowingPlantHeadBlock extends GrowingPlantBlock implements
     protected GrowingPlantHeadBlock getHeadBlock() {
         return this;
     }
+
+    // Leaves start - zero tick plants
+    @Override
+    protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (!state.canSurvive(level, pos)) {
+            level.destroyBlock(pos, true);
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        }
+    }
+    // Leaves end - zero tick plants
 }
diff --git a/net/minecraft/world/level/block/SugarCaneBlock.java b/net/minecraft/world/level/block/SugarCaneBlock.java
index 373a61d2d8e73f3bbfd5117c492a8d63bb574f54..629b8ef749543eb3be9756ef917b0bbb4537e723 100644
--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -43,6 +43,10 @@ public class SugarCaneBlock extends Block {
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
+        // Leaves start - zero tick plants
+        } else if (org.leavesmc.leaves.LeavesConfig.modify.oldMC.zeroTickPlants) {
+            this.randomTick(state, level, pos, random);
+        // Leaves end - zero tick plants
         }
     }
 
