From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Mon, 30 Oct 2023 10:43:44 +0800
Subject: [PATCH] Make Item tick vanilla


diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index bb2308339e614de8753b69a0961b35e4bbbc5d96..87e87fb2b11cacb69170a245b70578c71fef4b9b 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -134,6 +134,9 @@ public class ItemEntity extends Entity implements TraceableEntity {
     // Paper start - EAR 2
     @Override
     public void inactiveTick() {
+        // Leaves start - vanilla
+        this.tick();
+        /*
         super.inactiveTick();
         if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
             this.pickupDelay--;
@@ -151,6 +154,8 @@ public class ItemEntity extends Entity implements TraceableEntity {
             // CraftBukkit end
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
+         */
+        // Leaves end - vanilla
     }
     // Paper end - EAR 2
 
