From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Sun, 26 Jul 2026 11:22:25 +0800
Subject: [PATCH] Cache item max stack size


diff --git a/net/minecraft/core/component/PatchedDataComponentMap.java b/net/minecraft/core/component/PatchedDataComponentMap.java
index 1c9218686c50d59200c53d164a33debc41b1de8b..6ae08f2175dee9d88b79eb67a67b57b1ae39f770 100644
--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -29,6 +29,7 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         this.prototype = prototype;
         this.patch = patch;
         this.copyOnWrite = copyOnWrite;
+        recalculateMaxStackSize(null); // Leaves - cache item max stack size
     }
 
     public static PatchedDataComponentMap fromPatch(final DataComponentMap prototype, final DataComponentPatch patch) {
@@ -75,7 +76,7 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         } else {
             lastValue = (Optional<T>)this.patch.put(type, Optional.ofNullable(value));
         }
-
+        recalculateMaxStackSize(type); // Leaves - cache item max stack size
         return lastValue != null ? lastValue.orElse(defaultValue) : defaultValue;
     }
 
@@ -92,7 +93,7 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         } else {
             lastValue = (Optional<? extends T>)this.patch.remove(type);
         }
-
+        recalculateMaxStackSize(type); // Leaves - cache item max stack size
         return (T)(lastValue != null ? lastValue.orElse(null) : defaultValue);
     }
 
@@ -102,6 +103,7 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         for (Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(patch.map)) {
             this.applyPatch(entry.getKey(), entry.getValue());
         }
+        recalculateMaxStackSize(null); // Leaves -
     }
 
     private void applyPatch(final DataComponentType<?> type, final Optional<?> value) {
@@ -117,23 +119,27 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         } else {
             this.patch.remove(type);
         }
+        recalculateMaxStackSize(null); // Leaves - cache item max stack size
     }
 
     public void restorePatch(final DataComponentPatch patch) {
         this.ensureMapOwnership();
         this.patch.clear();
         this.patch.putAll(patch.map);
+        recalculateMaxStackSize(null); // Leaves - cache item max stack size
     }
 
     public void clearPatch() {
         this.ensureMapOwnership();
         this.patch.clear();
+        recalculateMaxStackSize(null); // Leaves - cache item max stack size
     }
 
     public void setAll(final DataComponentMap components) {
         for (TypedDataComponent<?> entry : components) {
             entry.applyTo(this);
         }
+        recalculateMaxStackSize(null); // Leaves - cache item max stack size
     }
 
     private void ensureMapOwnership() {
@@ -252,4 +258,19 @@ public final class PatchedDataComponentMap implements DataComponentMap, org.leav
         return 0;
     }
     // Leaves end - Lithium Sleeping Block Entity
+
+    // Leaves start - cache item max stack size
+    private int cachedMaxStackSize = 1;
+
+    public int getMaxStackSize() {
+        return cachedMaxStackSize;
+    }
+
+    private <T> void recalculateMaxStackSize(@Nullable DataComponentType<T> type) {
+        if (type != null && !type.equals(net.minecraft.core.component.DataComponents.MAX_STACK_SIZE)) {
+            return;
+        }
+        cachedMaxStackSize = getOrDefault(net.minecraft.core.component.DataComponents.MAX_STACK_SIZE, 1);
+    }
+    // Leaves end - cache item max stack size
 }
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 2cb3347dbad74c19fd1fe1b4a4a28ab354fad555..4285aed698859310c14714d7b5eb5782c0edef62 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -1507,4 +1507,11 @@ public final class ItemStack implements DataComponentHolder, ItemInstance, Chang
         }
     }
     // Leaves end - Lithium Sleeping Block Entity
+
+    // Leaves start - cache item max stack size
+    @Override
+    public int getMaxStackSize() {
+        return components.getMaxStackSize();
+    }
+    // Leaves end - cache item max stack size
 }
