From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Thu, 6 Feb 2025 00:14:23 +0800
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 777214ea32b1781e6c53b83714a6d141b3aa2b37..c9b9ee761457d7186dad12d72d976dde4bc7210c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -416,6 +416,8 @@ public final class CraftServer implements Server {
 
         CraftRegistry.setMinecraftRegistry(console.registryAccess());
 
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.init(); // Leaves - Bytebuf API
+
         if (!Main.useConsole) {
             this.getLogger().info("Console input is disabled due to --noconsole command argument");
         }
@@ -2982,4 +2984,11 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Leaves end - replay mod api
+
+    // Leaves start - Bytebuf API
+    @Override
+    public org.leavesmc.leaves.bytebuf.BytebufAllocator getBytebufAllocator() {
+        return org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.allocator();
+    }
+    // Leaves end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ea42e212c9114635316ac6109347b563b9cd7349..5b8bd3c5fff7f1b7564f2ddd80f38131ab88ee0e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3364,4 +3364,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         super.knockback(strength, directionX, directionZ);
         this.entity.hurtMarked = true;
     }
+
+    // Leaves start - Bytebuf API
+    @Override
+    public void send(org.leavesmc.leaves.bytebuf.PacketType packetType, org.leavesmc.leaves.bytebuf.Bytebuf bytebuf) {
+        org.leavesmc.leaves.bytebuf.internal.InternalBytebufHandler.sendPacket(this, packetType, bytebuf);
+    }
+
+    @Override
+    public @org.jetbrains.annotations.NotNull Object getChannel() {
+        return getHandle().connection.connection.channel;
+    }
+    // Leaves end - Bytebuf API
 }
